### PR TITLE
Improve typescript typings and jsdocs

### PR DESCRIPTION
**Pull request by loban**
Makes the code more explicit and easier to understand.

### CODE REVIEW

File: src/AccessControl.ts [hunk 1]
```diff
@@ -206,42 +206,41 @@ class AccessControl {
...
     }
     /**
      * Get roles which allow this permission
-     * @param {IQueryInfo} - permission query object we want to check
+     * @param {IQueryInfo} query - permission query object we want to check
      *
      * @returns {String[]} - roles
      */
-    allowingRoles(query: IQueryInfo) {
+    async allowingRoles(query: IQueryInfo): Promise<string[]> {
```
[the remark this answers is not in the document]
Comment (loban) [reply src/AccessControl.ts]: False positive?

File: src/AccessControl.ts [hunk 2]
```diff
@@ -206,42 +206,41 @@ class AccessControl {
...
 
     /**
      * Get allowed actions of resource when conditions are skipped
-     * @param {IQueryInfo} - permission query object we want to check
+     * @param {IQueryInfo} query - permission query object we want to check
      *
      *  @returns {String[]} - actions
      */
-    allowedActions(query: IQueryInfo) {
+    async allowedActions(query: IQueryInfo): Promise<string[]> {
```
[the remark this answers is not in the document]
Comment (loban) [reply src/AccessControl.ts]: False positive?

File: src/AccessControl.ts [hunk 3]
```diff
@@ -206,42 +206,41 @@ class AccessControl {
...
 
     /**
      * Get allowed resources when conditions are skipped
-     * @param {IQueryInfo} - permission query object we want to check
+     * @param {IQueryInfo} query - permission query object we want to check
      *
      *  @returns {String[]} - resources
      */
-    allowedResources(query: IQueryInfo) {
+    async allowedResources(query: IQueryInfo): Promise<string[]> {
```
[the remark this answers is not in the document]
Comment (loban) [reply src/AccessControl.ts]: False positive?

File: src/utils/common.ts [hunk 4]
```diff
@@ -350,12 +350,12 @@ export class CommonUtil {
         });
     }
 
-    public static matchesAllElement(values: any, predicateFn: (elm) => boolean) {
+    public static matchesAllElement(values: any, predicateFn: (elm) => boolean): boolean {
```
[the remark this answers is not in the document]
Comment (loban) [reply src/utils/common.ts]: False positive?

File: src/utils/common.ts [hunk 5]
```diff
@@ -350,12 +350,12 @@ export class CommonUtil {
...
         values = ArrayUtil.toArray(values);
         return values.every(predicateFn);
     }
 
-    public static matchesAnyElement(values: any, predicateFn: (elm) => boolean) {
+    public static matchesAnyElement(values: any, predicateFn: (elm) => boolean): boolean {
```
[the remark this answers is not in the document]
Comment (loban) [reply src/utils/common.ts]: False positive?

**Comment by koladilip** [comment]
I have published this version as https://www.npmjs.com/package/role-acl/v/4.0.6

**Pull request merged**